include/libdnf/base/base.hpp

```diff
@@ -92,6 +93,18 @@ class Base {
     void add_plugin(plugin::IPlugin & iplugin_instance);
     void load_plugins();
 
+    /// Set DownloadCallbacks factory class that will be used to create callbacks for
+    /// all downloads.
+    /// @param download_callbacks_factory   Instance of implementation of DownloadCallbacksFactory interface
+    void set_download_callbacks_factory(
+        std::unique_ptr<libdnf::repo::DownloadCallbacksFactory> && download_callbacks_factory);
+
+    /// Create DownloadCallbacks instance using the DownloadCallbacksFactory
+    /// (see set_download_callbacks_factory() method). In case the factory is
+    /// not set, nullptr is returned.
+    /// @param what String that describes what is being downloaded (URL, package...)
+    std::unique_ptr<libdnf::repo::DownloadCallbacks> create_download_callbacks(const std::string & what);
```

Comment:
Do we need it in public part? It looks like we use it only internally.

Reply:
Right. I was just reluctant to add another two friend classes to Base. But I guess it's better to hide as much functionality as we can. I'll move it.